File: apps/kotlin/src/test/kotlin/chapter08/SurvivedClassifierTest.kt
package chapter08

import chapter02.TrainTestSplit
import chapter02.splitTrainTest
import chapter03.DecisionTree
import chapter03.gini
import chapter03.predictWithTribuo
import chapter03.trainTribuoTree
import dataset.dataDir
import org.jetbrains.kotlinx.dataframe.AnyFrame
import org.jetbrains.kotlinx.dataframe.api.asColumnGroup
import org.jetbrains.kotlinx.dataframe.api.dataFrameOf
import org.jetbrains.kotlinx.dataframe.api.pivotMatches
import org.jetbrains.kotlinx.dataframe.api.toMap
import org.junit.jupiter.api.Assumptions.assumeTrue
import support.captureStdout
import java.io.File
import java.io.InvalidClassException
import java.io.ObjectOutputStream
import java.nio.file.Path
import kotlin.io.path.createTempDirectory
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNull
import kotlin.test.assertTrue

private const val HEADER = "\uFEFFPassengerId,Survived,Pclass,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked\n"

class LoadSurvivedTest {
    private val directory: Path = createTempDirectory()

    private fun writeCsv(rows: String): File = File(directory.toFile(), "survived.csv").apply { writeText(HEADER + rows) }

    @Test
    fun `BOM付きCSVを読み込み空欄を欠損値にする`() {
        val df = loadSurvived(writeCsv("1,0,3,male,,0,0,X-1,8.5,,S\n"))

        assertEquals("male", df["Sex"][0])
        assertNull(df["Age"][0])
        assertNull(df["Cabin"][0])
    }

    @Test
    fun `1文字の値の列も文字列として読み込む`() {
        val df = loadSurvived(writeCsv("1,0,3,male,,0,0,X-1,8.5,,S\n"))

        assertEquals("S", df["Embarked"][0])
    }
}

class GroupMedianImputerTest {
    @Test
    fun `同じグループの中央値で欠損値を補完する`() {
        val df =
            dataFrameOf(
                "Pclass" to listOf(1, 1, 1, 1),
                "Sex" to listOf("female", "female", "female", "female"),
                "Age" to listOf(20.0, 30.0, 70.0, null),
            )
        val imputer = GroupMedianImputer(column = "Age", by = listOf("Pclass", "Sex"))

        val filled = imputer.fit(df).transform(df)

        assertEquals(listOf(20.0, 30.0, 70.0, 30.0), filled["Age"].toList())
    }

    @Test
    fun `グループごとに異なる中央値で補完する`() {
        val df =
            dataFrameOf(
                "Pclass" to listOf(1, 1, 1, 3, 3, 3),
                "Sex" to listOf("female", "female", "female", "male", "male", "male"),
                "Age" to listOf(40.0, 50.0, null, 10.0, 20.0, null),
            )
        val imputer = GroupMedianImputer(column = "Age", by = listOf("Pclass", "Sex"))

        val filled = imputer.fit(df).transform(df)

        assertEquals(listOf(40.0, 50.0, 45.0, 10.0, 20.0, 15.0), filled["Age"].toList())
    }

    @Test
    fun `訓練データで求めた中央値を別のデータの補完に使う`() {
        val train = dataFrameOf("Pclass" to listOf(2, 2), "Sex" to listOf("male", "male"), "Age" to listOf(30.0, 34.0))
        val other = dataFrameOf("Pclass" to listOf(2), "Sex" to listOf("male"), "Age" to listOf(null))
        val imputer = GroupMedianImputer(column = "Age", by = listOf("Pclass", "Sex"))

        val filled = imputer.fit(train).transform(other)

        assertEquals(listOf(32.0), filled["Age"].toList())
    }

    @Test
    fun `訓練データに無いグループは全体の中央値で補完する`() {
        val train =
            dataFrameOf(
                "Pclass" to listOf(1, 1, 3),
                "Sex" to listOf("female", "female", "male"),
                "Age" to listOf(30.0, 40.0, 20.0),
            )
        val other = dataFrameOf("Pclass" to listOf(2), "Sex" to listOf("female"), "Age" to listOf(null))
        val imputer = GroupMedianImputer(column = "Age", by = listOf("Pclass", "Sex"))

        val filled = imputer.fit(train).transform(other)

        assertEquals(listOf(30.0), filled["Age"].toList())
    }

    @Test
    fun `元のデータフレームは変更しない`() {
        val df = dataFrameOf("Pclass" to listOf(1, 1), "Sex" to listOf("male", "male"), "Age" to listOf(30.0, null))
        val imputer = GroupMedianImputer(column = "Age", by = listOf("Pclass", "Sex"))

        imputer.fit(df).transform(df)

        assertNull(df["Age"][1])
    }

    @Test
    fun `年齢がすべて欠けたグループは全体の中央値で補完する`() {
        val df =
            dataFrameOf(
                "Pclass" to listOf(1, 1, 2),
                "Sex" to listOf("female", "female", "female"),
                "Age" to listOf(30.0, 40.0, null),
            )
        val imputer = GroupMedianImputer(column = "Age", by = listOf("Pclass", "Sex"))

        val filled = imputer.fit(df).transform(df)

        assertEquals(listOf(30.0, 40.0, 35.0), filled["Age"].toList())
    }
}

class MostFrequentImputerTest {
    @Test
    fun `訓練データで最も多い値で欠損値を補完する`() {
        val train = dataFrameOf("Embarked" to listOf("S", "C", "S", null))
        val other = dataFrameOf("Embarked" to listOf(null, "Q"))
        val imputer = MostFrequentImputer(column = "Embarked")

        val filled = imputer.fit(train).transform(other)

        assertEquals(listOf("S", "Q"), filled["Embarked"].toList())
    }

    @Test
    fun `欠損値だけの列も補完する`() {
        val train = dataFrameOf("Embarked" to listOf("S", "C", "S"))
        val other = dataFrameOf("Embarked" to listOf(null))
        val imputer = MostFrequentImputer(column = "Embarked")

        val filled = imputer.fit(train).transform(other)

        assertEquals(listOf("S"), filled["Embarked"].toList())
    }
}

class PivotMatchesLearningTest {
    @Test
    fun `pivotMatchesはデータに含まれるカテゴリの列しか作らない`() {
        val train = dataFrameOf("PassengerId" to listOf(1, 2, 3), "Embarked" to listOf("C", "Q", "S"))
        val other = dataFrameOf("PassengerId" to listOf(4), "Embarked" to listOf("S"))

        assertEquals(listOf("C", "Q", "S"), train.pivotMatches("Embarked")["Embarked"].asColumnGroup().columnNames())
        assertEquals(listOf("S"), other.pivotMatches("Embarked")["Embarked"].asColumnGroup().columnNames())
    }
}

class DummyEncoderTest {
    @Test
    fun `2値のカテゴリを最初のカテゴリを除いた0と1の列にする`() {
        val df = dataFrameOf("Pclass" to listOf(1, 3, 2), "Sex" to listOf("female", "male", "male"))
        val encoder = DummyEncoder(columns = listOf("Sex"))

        val encoded = encoder.fit(df).transform(df)

        assertEquals(mapOf("Pclass" to listOf(1, 3, 2), "Sex_male" to listOf(0, 1, 1)), encoded.toMap())
    }

    @Test
    fun `別のデータにも訓練データと同じ列を作る`() {
        val train = dataFrameOf("Embarked" to listOf("C", "Q", "S"))
        val other = dataFrameOf("Embarked" to listOf("S", "S"))
        val encoder = DummyEncoder(columns = listOf("Embarked"))

        val encoded = encoder.fit(train).transform(other)

        assertEquals(mapOf("Embarked_Q" to listOf(0, 0), "Embarked_S" to listOf(1, 1)), encoded.toMap())
    }
}

class WeightedGiniTest {
    @Test
    fun `重みがすべて1なら第3章のジニ不純度と同じ値になる`() {
        val labels = listOf(0, 1, 1)

        assertEquals(gini(labels.map { it.toString() }), weightedGini(labels, weights = listOf(1.0, 1.0, 1.0)))
    }

    @Test
    fun `重みの大きいラベルほど多いものとして不純度を計算する`() {
        val impurity = weightedGini(listOf(0, 1), weights = listOf(1.0, 3.0))

        assertEquals(0.375, impurity, absoluteTolerance = 1e-12)
    }
}

class BalancedWeightsTest {
    @Test
    fun `少ないクラスほど1件の重みを大きくしクラスごとの重みの合計をそろえる`() {
        val weights = balancedWeights(listOf(0, 0, 0, 1))

        assertEquals(listOf(4.0 / 6, 4.0 / 6, 4.0 / 6, 2.0), weights)
        assertEquals(weights.take(3).sum(), weights[3], absoluteTolerance = 1e-12)
    }
}

class DecisionTreeClassifierTest {
    private val x =
        dataFrameOf(
            "Fare" to listOf(8.0, 9.0, 13.0, 20.0, 60.0, 80.0),
            "Age" to listOf(30.0, 22.0, 18.0, 45.0, 25.0, 33.0),
        )
    private val t = listOf(0, 0, 1, 0, 1, 1)

    @Test
    fun `重み付けなしなら第3章の決定木と同じ予測をする`() {
        for (maxDepth in listOf(1, 2, null)) {
            val expected = DecisionTree(maxDepth).fit(x, t.map { it.toString() }).predict(x)

            val predictions = DecisionTreeClassifier(maxDepth, ClassWeight.NONE).fit(x, t).predict(x)

            assertEquals(expected, predictions.map { it.toString() }, "深さ $maxDepth")
        }
    }

    @Test
    fun `balancedにすると少ないクラスが混ざった葉でも少ないクラスを予測する`() {
        val fare = dataFrameOf("Fare" to listOf(1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0))
        val survived = listOf(0, 0, 0, 0, 0, 0, 1)
        val newX = dataFrameOf("Fare" to listOf(1.0, 2.0))

        val none = DecisionTreeClassifier(maxDepth = 1, classWeight = ClassWeight.NONE).fit(fare, survived)
        val balanced = DecisionTreeClassifier(maxDepth = 1, classWeight = ClassWeight.BALANCED).fit(fare, survived)

        assertEquals(listOf(0, 0), none.predict(newX))
        assertEquals(listOf(0, 1), balanced.predict(newX))
    }
}

class SplitFeaturesAndTargetTest {
    @Test
    fun `特徴量の列とSurvived列に分ける`() {
        val df =
            dataFrameOf(
                "PassengerId" to listOf(1),
                "Survived" to listOf(1),
                "Pclass" to listOf(2),
                "Sex" to listOf("female"),
                "Age" to listOf(28.0),
                "SibSp" to listOf(0),
                "Parch" to listOf(1),
                "Ticket" to listOf("X-2"),
                "Fare" to listOf(15.0),
                "Cabin" to listOf(null),
                "Embarked" to listOf("C"),
            )

        val (x, t) = splitFeaturesAndTarget(df)

        assertEquals(listOf("Pclass", "Sex", "Age", "SibSp", "Parch", "Fare", "Embarked"), x.columnNames())
        assertEquals(listOf(1), t)
    }
}

private fun passengers(vararg rows: List<Any?>): AnyFrame =
    dataFrameOf(
        *FEATURES
            .mapIndexed { i, name ->
                name to
                    rows.map {
                        it[i]
                    }
            }.toTypedArray(),
    )

private fun trainingData(): Pair<AnyFrame, List<Int>> {
    val x =
        passengers(
            listOf(1, "female", 30.0, 0, 0, 80.0, "C"),
            listOf(2, "female", null, 1, 0, 20.0, "S"),
            listOf(3, "female", 22.0, 0, 1, 9.0, null),
            listOf(3, "female", 18.0, 0, 0, 8.0, "Q"),
            listOf(1, "male", 45.0, 0, 0, 60.0, "S"),
            listOf(2, "male", null, 0, 0, 13.0, "S"),
            listOf(3, "male", 25.0, 1, 0, 7.0, "S"),
            listOf(3, "male", 33.0, 0, 0, 8.0, null),
        )
    val t = listOf(1, 1, 1, 1, 0, 0, 0, 0)
    return x to t
}

private fun newPassengers(): AnyFrame =
    passengers(
        listOf(2, "female", null, 0, 0, 12.0, null),
        listOf(1, "male", null, 1, 1, 70.0, "C"),
    )

class BuildPipelineTest {
    @Test
    fun `欠損値を含むデータで学習して予測できる`() {
        val (x, t) = trainingData()
        val pipeline = buildPipeline(maxDepth = 3, classWeight = ClassWeight.NONE)

        val fitted = pipeline.fit(x, t)

        assertEquals(listOf(1, 0), fitted.predict(newPassengers()))
    }

    @Test
    fun `クラスの重みと深さをモデルに渡す`() {
        val pipeline = buildPipeline(maxDepth = 5, classWeight = ClassWeight.BALANCED)

        assertEquals(DecisionTreeClassifier(maxDepth = 5, classWeight = ClassWeight.BALANCED), pipeline.model)
    }
}

class SaveAndLoadModelTest {
    private val directory: Path = createTempDirectory()

    @Test
    fun `保存したパイプラインを読み込むと同じ予測をする`() {
        val (x, t) = trainingData()
        val pipeline = buildPipeline(maxDepth = 3, classWeight = ClassWeight.NONE).fit(x, t)
        val modelFile = File(directory.toFile(), "model/survived.ser")

        saveModel(pipeline, modelFile)
        val loaded = loadModel(modelFile)

        assertEquals(listOf(1, 0), loaded.predict(newPassengers()))
    }

    @Test
    fun `許可していないクラスを含むファイルは読み込まない`() {
        val modelFile = File(directory.toFile(), "unknown.ser")
        ObjectOutputStream(modelFile.outputStream()).use { it.writeObject(java.awt.Point(1, 2)) }

        assertFailsWith<InvalidClassException> { loadModel(modelFile) }
    }
}

class EvaluateTest {
    @Test
    fun `正解率と見つけた生存者の数を求める`() {
        val (x, t) = trainingData()
        val split = TrainTestSplit(xTrain = x, xTest = newPassengers(), tTrain = t, tTest = listOf(1, 1))
        val pipeline = buildPipeline(maxDepth = 3, classWeight = ClassWeight.NONE).fit(x, t)

        val evaluation = evaluate(pipeline, split)

        assertEquals(Evaluation(trainAccuracy = 1.0, testAccuracy = 0.5, foundSurvivors = 1, survivors = 2), evaluation)
    }
}

class SurvivedDataTest {
    private val csvFile = File(dataDir(), "Survived.csv")

    @BeforeTest
    fun requireData() {
        assumeTrue(csvFile.exists(), "学習データ Survived.csv が配置されていない（gulp data:setup）")
    }

    private fun survivedSplit(): TrainTestSplit<Int> {
        val (x, t) = splitFeaturesAndTarget(loadSurvived(csvFile))
        return splitTrainTest(x, t, testSize = 0.2, seed = 0)
    }

    private fun evaluateWith(
        split: TrainTestSplit<Int>,
        maxDepth: Int,
        classWeight: ClassWeight,
    ): Evaluation = evaluate(buildPipeline(maxDepth, classWeight).fit(split.xTrain, split.tTrain), split)

    @Test
    fun `実データの件数と欠損値の数を確認する`() {
        val df = loadSurvived(csvFile)

        assertEquals(891, df.rowsCount())
        assertEquals(
            mapOf("Age" to 177, "Cabin" to 687, "Embarked" to 2),
            listOf("Age", "Cabin", "Embarked").associateWith { name ->
                df[name].values().count {
                    it ==
                        null
                }
            },
        )
    }

    @Test
    fun `深さ2ではbalancedにすると見つけられる生存者が増える`() {
        val split = survivedSplit()

        assertEquals(33, evaluateWith(split, maxDepth = 2, classWeight = ClassWeight.NONE).foundSurvivors)
        assertEquals(51, evaluateWith(split, maxDepth = 2, classWeight = ClassWeight.BALANCED).foundSurvivors)
    }

    @Test
    fun `深さ5ではbalancedにしても見つけられる生存者は増えない`() {
        val split = survivedSplit()

        val none = evaluateWith(split, maxDepth = 5, classWeight = ClassWeight.NONE)
        val balanced = evaluateWith(split, maxDepth = 5, classWeight = ClassWeight.BALANCED)

        assertEquals(46 to 45, none.foundSurvivors to balanced.foundSurvivors)
        assertEquals(0.832, balanced.testAccuracy, absoluteTolerance = 1e-3)
    }

    @Test
    fun `深さ5までならTribuoの決定木と前処理後のテストデータの予測が一致する`() {
        val split = survivedSplit()

        for (maxDepth in 1..5) {
            val pipeline = buildPipeline(maxDepth, ClassWeight.NONE).fit(split.xTrain, split.tTrain)
            val xTrain = pipeline.transform(split.xTrain)
            val xTest = pipeline.transform(split.xTest)
            val tribuo = predictWithTribuo(trainTribuoTree(xTrain, split.tTrain.map { it.toString() }, maxDepth, 1.0f), xTest)

            assertEquals(pipeline.predict(split.xTest).map { it.toString() }, tribuo, "深さ $maxDepth")
        }
    }

    @Test
    fun `実行すると評価結果を表示してモデルを保存する`() {
        val modelFile = File(createTempDirectory().toFile(), "survived.ser")

        val output = captureStdout { main(modelFile) }

        assertTrue(modelFile.exists())
        assertEquals(
            "データ件数: 891（生存 342, 死亡 549）\n" +
                "訓練データ: 712 件, テストデータ: 179 件\n" +
                "classWeight=NONE: 訓練 0.851, テスト 0.832, 生存者 68 人中 46 人を発見\n" +
                "classWeight=BALANCED: 訓練 0.850, テスト 0.832, 生存者 68 人中 45 人を発見\n" +
                "保存したモデル: survived.ser\n" +
                "架空の乗客の予測: [1, 0]\n",
            output,
        )
    }

    @Test
    fun `重み付けなしなら深さ10まで第3章の決定木と前処理後のテストデータの予測が一致する`() {
        val split = survivedSplit()

        for (maxDepth in 1..10) {
            val pipeline = buildPipeline(maxDepth, ClassWeight.NONE).fit(split.xTrain, split.tTrain)
            val xTrain = pipeline.transform(split.xTrain)
            val xTest = pipeline.transform(split.xTest)
            val chapter03 = DecisionTree(maxDepth).fit(xTrain, split.tTrain.map { it.toString() }).predict(xTest)

            assertEquals(pipeline.predict(split.xTest).map { it.toString() }, chapter03, "深さ $maxDepth")
        }
    }
}

File: apps/node/test/chapter08/survived-classifier.test.ts
import { mkdtempSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { describe, expect, it } from "vitest";
import { DecisionTree, gini } from "../../src/chapter03/decision-tree.ts";
import {
  balancedWeights,
  fitDecisionTree,
  predictDecisionTree,
  weightedGini,
} from "../../src/chapter08/decision-tree-classifier.ts";
import { evaluate } from "../../src/chapter08/evaluation.ts";
import {
  fitPipeline,
  loadModel,
  predict,
  predictPassenger,
  saveModel,
} from "../../src/chapter08/pipeline.ts";
import {
  FEATURES,
  loadSurvived,
  type Passenger,
  splitFeaturesAndTarget,
} from "../../src/chapter08/survived-data.ts";
import {
  encodeDummies,
  fitDummyEncoder,
  fitGroupMedianImputer,
  fitMostFrequentImputer,
  imputeGroupMedian,
  imputeMostFrequent,
} from "../../src/chapter08/transformers.ts";

const HEADER =
  "\uFEFFPassengerId,Survived,Pclass,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked\n";

function writeCsv(rows: string): string {
  const directory = mkdtempSync(join(tmpdir(), "survived-"));
  const csvFile = join(directory, "survived.csv");
  writeFileSync(csvFile, HEADER + rows);
  return csvFile;
}

describe("loadSurvived", () => {
  it("BOM 付き CSV を読み込み、数値の列を数値に、空欄を null にする", () => {
    const rows = loadSurvived(writeCsv("1,0,3,male,,0,0,X-1,8.5,,S\n"));

    expect(rows[0]?.Pclass).toBe(3);
    expect(rows[0]?.Fare).toBe(8.5);
    expect(rows[0]?.Age).toBeNull();
    expect(rows[0]?.Cabin).toBeNull();
  });

  it("文字列の列は文字列のまま読み込む", () => {
    const rows = loadSurvived(writeCsv("1,0,3,male,,0,0,X-1,8.5,,S\n"));

    expect(rows[0]?.Sex).toBe("male");
    expect(rows[0]?.Embarked).toBe("S");
  });
});

describe("Map のキー（学習用テスト）", () => {
  it("配列のキーは要素の値ではなく、同じ配列かどうかで比べる", () => {
    const medians = new Map([[[1, "female"], 45]]);

    expect(medians.get([1, "female"])).toBeUndefined();
  });

  it("JSON の文字列にしたキーなら要素の値で引ける", () => {
    const medians = new Map([[JSON.stringify([1, "female"]), 45]]);

    expect(medians.get(JSON.stringify([1, "female"]))).toBe(45);
  });

  it("Map を JSON の文字列にすると中身が失われる", () => {
    const medians = new Map([['[1,"female"]', 45]]);

    expect(JSON.stringify({ medians })).toBe('{"medians":{}}');
  });
});

describe("GroupMedianImputer", () => {
  it("同じグループの中央値で欠損値を補完する", () => {
    const x = [
      { Pclass: 1, Sex: "female", Age: 20 },
      { Pclass: 1, Sex: "female", Age: 30 },
      { Pclass: 1, Sex: "female", Age: 70 },
      { Pclass: 1, Sex: "female", Age: null },
    ];

    const imputer = fitGroupMedianImputer(x, "Age", ["Pclass", "Sex"]);

    const filled = imputeGroupMedian(x, imputer);
    expect(filled.map((row) => row.Age)).toEqual([20, 30, 70, 30]);
  });

  it("グループごとに異なる中央値で補完する", () => {
    const x = [
      { Pclass: 1, Sex: "female", Age: 40 },
      { Pclass: 1, Sex: "female", Age: 50 },
      { Pclass: 1, Sex: "female", Age: null },
      { Pclass: 3, Sex: "male", Age: 10 },
      { Pclass: 3, Sex: "male", Age: 20 },
      { Pclass: 3, Sex: "male", Age: null },
    ];

    const imputer = fitGroupMedianImputer(x, "Age", ["Pclass", "Sex"]);

    const filled = imputeGroupMedian(x, imputer);
    expect(filled.map((row) => row.Age)).toEqual([40, 50, 45, 10, 20, 15]);
  });

  it("訓練データで求めた中央値を別のデータの補完に使う", () => {
    const train = [
      { Pclass: 2, Sex: "male", Age: 30 },
      { Pclass: 2, Sex: "male", Age: 34 },
    ];
    const other = [{ Pclass: 2, Sex: "male", Age: null }];

    const imputer = fitGroupMedianImputer(train, "Age", ["Pclass", "Sex"]);

    expect(imputeGroupMedian(other, imputer)[0]?.Age).toBe(32);
  });

  it("訓練データに無いグループは全体の中央値で補完する", () => {
    const train = [
      { Pclass: 1, Sex: "female", Age: 30 },
      { Pclass: 1, Sex: "female", Age: 40 },
      { Pclass: 3, Sex: "male", Age: 20 },
    ];
    const other = [{ Pclass: 2, Sex: "female", Age: null }];

    const imputer = fitGroupMedianImputer(train, "Age", ["Pclass", "Sex"]);

    expect(imputeGroupMedian(other, imputer)[0]?.Age).toBe(30);
  });

  it("年齢がすべて欠けたグループは全体の中央値で補完する", () => {
    const x = [
      { Pclass: 1, Sex: "female", Age: 30 },
      { Pclass: 1, Sex: "female", Age: 40 },
      { Pclass: 2, Sex: "female", Age: null },
    ];

    const imputer = fitGroupMedianImputer(x, "Age", ["Pclass", "Sex"]);

    expect(imputeGroupMedian(x, imputer)[2]?.Age).toBe(35);
  });

  it("元のデータは変更しない", () => {
    const x = [
      { Pclass: 1, Sex: "male", Age: 30 },
      { Pclass: 1, Sex: "male", Age: null },
    ];

    imputeGroupMedian(x, fitGroupMedianImputer(x, "Age", ["Pclass", "Sex"]));

    expect(x[1]?.Age).toBeNull();
  });
});

describe("MostFrequentImputer", () => {
  it("訓練データで最も多い値で欠損値を補完する", () => {
    const train = [
      { Embarked: "S" },
      { Embarked: "C" },
      { Embarked: "S" },
      { Embarked: null },
    ];
    const other = [{ Embarked: null }, { Embarked: "Q" }];

    const imputer = fitMostFrequentImputer(train, "Embarked");

    const filled = imputeMostFrequent(other, imputer);
    expect(filled.map((row) => row.Embarked)).toEqual(["S", "Q"]);
  });
});

describe("DummyEncoder", () => {
  it("2 値のカテゴリを、最初のカテゴリを除いた 0 と 1 の列にする", () => {
    const x = [
      { Pclass: 1, Sex: "female" },
      { Pclass: 3, Sex: "male" },
      { Pclass: 2, Sex: "male" },
    ];

    const encoder = fitDummyEncoder(x, ["Sex"]);

    expect(encodeDummies(x, encoder)).toEqual([
      { Pclass: 1, Sex_male: 0 },
      { Pclass: 3, Sex_male: 1 },
      { Pclass: 2, Sex_male: 1 },
    ]);
  });

  it("別のデータにも訓練データと同じ列を作る", () => {
    const train = [{ Embarked: "C" }, { Embarked: "Q" }, { Embarked: "S" }];
    const other = [{ Embarked: "S" }, { Embarked: "S" }];

    const encoder = fitDummyEncoder(train, ["Embarked"]);

    expect(encodeDummies(other, encoder)).toEqual([
      { Embarked_Q: 0, Embarked_S: 1 },
      { Embarked_Q: 0, Embarked_S: 1 },
    ]);
  });
});

describe("weightedGini", () => {
  it("重みがすべて 1 なら第 3 章のジニ不純度と同じ値になる", () => {
    const labels = [0, 1, 1];

    expect(weightedGini(labels, [1, 1, 1])).toBe(gini(labels.map(String)));
  });

  it("重みの大きいラベルほど多いものとして不純度を計算する", () => {
    expect(weightedGini([0, 1], [1, 3])).toBeCloseTo(0.375, 12);
  });
});

describe("balancedWeights", () => {
  it("少ないクラスほど 1 件の重みを大きくし、クラスごとの重みの合計をそろえる", () => {
    const weights = balancedWeights([0, 0, 0, 1]);

    expect(weights).toEqual([4 / 6, 4 / 6, 4 / 6, 2]);
    expect(weights[3]).toBeCloseTo(3 * (4 / 6), 12);
  });
});

describe("fitDecisionTree", () => {
  const x = [
    { Fare: 8, Age: 30 },
    { Fare: 9, Age: 22 },
    { Fare: 13, Age: 18 },
    { Fare: 20, Age: 45 },
    { Fare: 60, Age: 25 },
    { Fare: 80, Age: 33 },
  ];
  const t = [0, 0, 1, 0, 1, 1];

  it.each([1, 2, undefined])(
    "重み付けなしなら第 3 章の決定木と同じ予測をする（深さ %s）",
    (maxDepth) => {
      const expected = new DecisionTree({ maxDepth })
        .fit(x, t.map(String))
        .predict(x);

      const tree = fitDecisionTree(x, t, { maxDepth, classWeight: "none" });

      expect(predictDecisionTree(tree, x).map(String)).toEqual(expected);
    },
  );

  it("balanced にすると少ないクラスが混ざった葉でも少ないクラスを予測する", () => {
    const fares = [1, 1, 1, 1, 2, 2, 2].map((Fare) => ({ Fare }));
    const survived = [0, 0, 0, 0, 0, 0, 1];
    const newX = [{ Fare: 1 }, { Fare: 2 }];

    const none = fitDecisionTree(fares, survived, {
      maxDepth: 1,
      classWeight: "none",
    });
    const balanced = fitDecisionTree(fares, survived, {
      maxDepth: 1,
      classWeight: "balanced",
    });

    expect(predictDecisionTree(none, newX)).toEqual([0, 0]);
    expect(predictDecisionTree(balanced, newX)).toEqual([0, 1]);
  });
});

describe("splitFeaturesAndTarget", () => {
  it("特徴量の列と Survived 列に分ける", () => {
    const rows = [
      {
        PassengerId: 1,
        Survived: 1,
        Pclass: 2,
        Sex: "female",
        Age: 28,
        SibSp: 0,
        Parch: 1,
        Ticket: "X-2",
        Fare: 15,
        Cabin: null,
        Embarked: "C",
      },
    ];

    const { x, t } = splitFeaturesAndTarget(rows);

    expect(Object.keys(x[0] ?? {})).toEqual([...FEATURES]);
    expect(t).toEqual([1]);
  });
});

type PassengerValues = [
  Pclass: number,
  Sex: string,
  Age: number | null,
  SibSp: number,
  Parch: number,
  Fare: number,
  Embarked: string | null,
];

function passengers(...rows: PassengerValues[]): Passenger[] {
  return rows.map(([Pclass, Sex, Age, SibSp, Parch, Fare, Embarked]) => ({
    Pclass,
    Sex,
    Age,
    SibSp,
    Parch,
    Fare,
    Embarked,
  }));
}

function trainingData(): { x: Passenger[]; t: number[] } {
  const x = passengers(
    [1, "female", 30, 0, 0, 80, "C"],
    [2, "female", null, 1, 0, 20, "S"],
    [3, "female", 22, 0, 1, 9, null],
    [3, "female", 18, 0, 0, 8, "Q"],
    [1, "male", 45, 0, 0, 60, "S"],
    [2, "male", null, 0, 0, 13, "S"],
    [3, "male", 25, 1, 0, 7, "S"],
    [3, "male", 33, 0, 0, 8, null],
  );
  return { x, t: [1, 1, 1, 1, 0, 0, 0, 0] };
}

function newPassengers(): Passenger[] {
  return passengers(
    [2, "female", null, 0, 0, 12, null],
    [1, "male", null, 1, 1, 70, "C"],
  );
}

describe("fitPipeline", () => {
  it("欠損値を含むデータで学習して予測できる", () => {
    const { x, t } = trainingData();

    const pipeline = fitPipeline(x, t, { maxDepth: 3, classWeight: "none" });

    expect(predict(pipeline, newPassengers())).toEqual([1, 0]);
  });
});

describe("saveModel と loadModel", () => {
  it("保存したパイプラインを読み込むと同じ予測をする", () => {
    const { x, t } = trainingData();
    const pipeline = fitPipeline(x, t, { maxDepth: 3, classWeight: "none" });
    const directory = mkdtempSync(join(tmpdir(), "model-"));
    const modelFile = join(directory, "model", "survived.json");

    saveModel(pipeline, modelFile);
    const loaded = loadModel(modelFile);

    expect(predict(loaded, newPassengers())).toEqual([1, 0]);
  });

  it("パイプラインの形をしていないファイルは読み込まない", () => {
    const directory = mkdtempSync(join(tmpdir(), "model-"));
    const modelFile = join(directory, "unknown.json");
    writeFileSync(modelFile, JSON.stringify({ tree: { kind: "leaf" } }));

    expect(() => loadModel(modelFile)).toThrow();
  });
});

describe("predictPassenger", () => {
  it("乗客 1 人分のデータから、生存（1）か死亡（0）かを予測する", () => {
    const { x, t } = trainingData();
    const pipeline = fitPipeline(x, t, { maxDepth: 3, classWeight: "none" });
    const [female, male] = newPassengers() as [Passenger, Passenger];

    expect(predictPassenger(pipeline, female)).toBe(1);
    expect(predictPassenger(pipeline, male)).toBe(0);
  });
});

describe("evaluate", () => {
  it("正解率と見つけた生存者の数を求める", () => {
    const { x, t } = trainingData();
    const split = {
      xTrain: x,
      xTest: newPassengers(),
      tTrain: t,
      tTest: [1, 1],
    };
    const pipeline = fitPipeline(x, t, { maxDepth: 3, classWeight: "none" });

    expect(evaluate(pipeline, split)).toEqual({
      trainAccuracy: 1,
      testAccuracy: 0.5,
      foundSurvivors: 1,
      survivors: 2,
    });
  });
});
